--- src/summaryview.c.summaryfix	Wed Aug 15 20:09:31 2001
+++ src/summaryview.c	Wed Aug 15 21:07:42 2001
@@ -136,6 +136,9 @@
 static void summary_set_menu_sensitive	(SummaryView		*summaryview);
 static void summary_set_add_sender_menu	(SummaryView		*summaryview);
 
+static void summary_select_node		(SummaryView		*summaryview,
+					 GtkCTreeNode		*node,
+					 gboolean		 display_msg);
 static guint summary_get_msgnum		(SummaryView		*summaryview,
 					 GtkCTreeNode		*node);
 
@@ -683,20 +686,22 @@
 		if (!summaryview->displayed)
 			messageview_clear(summaryview->messageview);
 		summary_select_by_msgnum(summaryview, selected_msgnum);
+		if (!summaryview->selected) {
+			/* no selected message - select first unread
+                           message, but do not display it */
+			node = summary_find_next_unread_msg(summaryview, NULL);
+			if (node == NULL && GTK_CLIST(ctree)->row_list != NULL)
+				node = GTK_CTREE_NODE
+					(GTK_CLIST(ctree)->row_list_end);
+			summary_select_node(summaryview, node, FALSE);
+		}
 	} else {
 		/* select first unread message */
 		node = summary_find_next_unread_msg(summaryview, NULL);
 		if (node == NULL && GTK_CLIST(ctree)->row_list != NULL)
 			node = GTK_CTREE_NODE(GTK_CLIST(ctree)->row_list_end);
-		if (node) {
-			GTK_EVENTS_FLUSH();
-			gtkut_ctree_expand_parent_all(ctree, node);
-			gtk_ctree_node_moveto(ctree, node, -1, 0.5, 0);
-			gtk_widget_grab_focus(GTK_WIDGET(ctree));
-			if (prefs_common.open_unread_on_enter)
-				summaryview->display_msg = TRUE;
-			gtk_sctree_select(GTK_SCTREE(ctree), node);
-		}
+		summary_select_node(summaryview, node,
+				    prefs_common.open_unread_on_enter);
 	}
 
 	summary_status_show(summaryview);
@@ -885,9 +890,26 @@
 void summary_select_by_msgnum(SummaryView *summaryview, guint msgnum)
 {
 	GtkCTreeNode *node;
-	GtkCTree *ctree = GTK_CTREE(summaryview->ctree);
 
 	node = summary_find_msg_by_msgnum(summaryview, msgnum);
+	summary_select_node(summaryview, node, FALSE);
+}
+
+/**
+ * summary_select_node:
+ * @summaryview: Summary view.
+ * @node: Summary tree node.
+ * @display_msg: TRUE to display the selected message.
+ * 
+ * Select @node (bringing it into view by scrolling and expanding its
+ * thread, if necessary) and unselect all others.  If @display_msg is
+ * TRUE, display the corresponding message in the message view.
+ **/
+static void summary_select_node(SummaryView *summaryview,
+				GtkCTreeNode *node,
+				gboolean display_msg)
+{
+	GtkCTree *ctree = GTK_CTREE(summaryview->ctree);
 
 	if (node) {
 		GTK_EVENTS_FLUSH();
@@ -895,7 +917,7 @@
 		gtk_ctree_node_moveto(ctree, node, -1, 0.5, 0);
 		gtk_widget_grab_focus(GTK_WIDGET(ctree));
 		gtk_sctree_unselect_all(GTK_SCTREE(ctree));
-		summaryview->display_msg = FALSE;
+		summaryview->display_msg = display_msg;
 		gtk_sctree_select(GTK_SCTREE(ctree), node);
 	}
 }
@@ -940,7 +962,7 @@
 
 	node = GTK_CTREE_NODE(GTK_CLIST(ctree)->row_list);
 
-	for (; node != NULL; node = GTK_CTREE_NODE_NEXT(node)) {
+	for (; node != NULL; node = gtkut_ctree_node_next(ctree, node)) {
 		msginfo = gtk_ctree_node_get_row_data(ctree, node);
 		if (msginfo->msgnum == msgnum) break;
 	}
