diff -u ../org/sylpheed-0.8.0/po/sylpheed.pot po/sylpheed.pot
--- ../org/sylpheed-0.8.0/src/mimeview.c	Fri May 17 19:45:49 2002
+++ src/mimeview.c	Fri Jul 19 09:58:11 2002
@@ -100,6 +100,7 @@

 static void mimeview_display_as_text	(MimeView	*mimeview);
 static void mimeview_save_as		(MimeView	*mimeview);
+static void mimeview_save_all		(MimeView	*mimeview);
 static void mimeview_launch		(MimeView	*mimeview);
 static void mimeview_open_with		(MimeView	*mimeview);
 static void mimeview_view_file		(const gchar	*filename,
@@ -114,7 +115,8 @@
 	{N_("/_Open"),		  NULL, mimeview_launch,	  0, NULL},
 	{N_("/Open _with..."),	  NULL, mimeview_open_with,	  0, NULL},
 	{N_("/_Display as text"), NULL, mimeview_display_as_text, 0, NULL},
-	{N_("/_Save as..."),	  NULL, mimeview_save_as,	  0, NULL}
+	{N_("/_Save as..."),	  NULL, mimeview_save_as,	  0, NULL},
+	{N_("/Save _all"),	  NULL, mimeview_save_all,	  0, NULL}
 #if USE_GPGME
         ,
         {N_("/_Check signature"), NULL, mimeview_check_signature, 0, NULL}
@@ -809,6 +811,56 @@
 	if (procmime_get_part(filename, mimeview->file, partinfo) < 0)
 		alertpanel_error
 			(_("Can't save the part of multipart message."));
+}
+
+static void mimeview_save_all(MimeView *mimeview)
+{
+	gchar *filename;
+	gchar *defname = NULL;
+	MimeInfo *partinfo;
+	MimeInfo *attachment;
+	gchar buf[1024];
+
+	if (!mimeview->opened) return;
+	if (!mimeview->file) return;
+
+	partinfo = gtk_ctree_node_get_row_data
+		(GTK_CTREE(mimeview->ctree), mimeview->opened);
+
+	g_return_if_fail(partinfo != NULL);
+
+	filename = filesel_select_file(_("Save as"), defname);
+
+	if (!filename) return;
+
+	// return to first children
+	attachment = partinfo->parent->children->next;
+
+	// for each attachment, extract it in the selected dir.
+	while(attachment != NULL)
+	{
+		if(attachment->filename)
+		{
+			g_snprintf(buf, sizeof(buf), "%s%s", filename, attachment->filename);
+
+			if (is_file_exist(buf))
+			{
+				AlertValue aval;
+
+				aval = alertpanel(_("Overwrite"),
+						_("Overwrite existing file?"),
+						_("OK"), _("Cancel"), NULL);
+
+				if (G_ALERTDEFAULT != aval) return;
+			}
+
+			if (procmime_get_part(buf, mimeview->file, attachment) < 0)
+				alertpanel_error
+					(_("Can't save the part of multipart message."));
+		}
+
+		attachment = attachment->next;
+	}
 }

 static void mimeview_launch(MimeView *mimeview)

diff -u ../org/sylpheed-0.8.0/po/sylpheed.pot po/sylpheed.pot
--- ../org/sylpheed-0.8.0/po/sylpheed.pot	Sun Jul 14 08:45:26 2002
+++ po/sylpheed.pot	Fri Jul 19 10:07:22 2002
@@ -3134,6 +3134,10 @@
 msgid "/_Save as..."
 msgstr ""

+#: src/mimeview.c:121
+msgid "/Save _all"
+msgstr ""
+
 #: src/mimeview.c:120
 msgid "/_Check signature"
 msgstr ""

diff -u ../org/sylpheed-0.8.0/po/fr.po po/fr.po
--- ../org/sylpheed-0.8.0/po/fr.po	Sun Jul 14 09:12:04 2002
+++ po/fr.po	Fri Jul 19 10:06:31 2002
@@ -3227,6 +3227,10 @@
 msgid "/_Save as..."
 msgstr "/Enregistrer so_us..."

+#: src/mimeview.c:121
+msgid "/Save _all"
+msgstr "/Détacher tout"
+
 #: src/mimeview.c:120
 msgid "/_Check signature"
 msgstr "/_Vérifier la signature"
