--- sylpheed-0.7.8/configure.in~	Sat Jun 15 18:04:23 2002
+++ sylpheed-0.7.8/configure.in	Sat Jun 15 18:05:47 2002
@@ -50,8 +50,6 @@
 AC_PROG_YACC
 AM_PROG_LIBTOOL
 
-SYLPHEED_ACLOCAL_INCLUDE(ac)
-
 case "$target" in
 *-darwin*)
 	CFLAGS="$CFLAGS -traditional-cpp -fno-common"
@@ -71,10 +69,10 @@
 AC_SUBST(faqdir)
 
 dnl Checks for libraries.
-AM_PATH_GLIB(1.2.6,,
+AM_PATH_GLIB(1.2.6,[],
 	AC_MSG_ERROR(Test for GLIB failed. See the file 'INSTALL' for help.),
 	gthread)
-AM_PATH_GTK(1.2.6,,
+AM_PATH_GTK(1.2.6,[],
 	AC_MSG_ERROR(Test for GTK failed. See the file 'INSTALL' for help.))
 
 dnl Check for built-in image view support
@@ -185,7 +183,7 @@
 	[  --disable-compface      Do not use compface (X-Face)],
 	[ac_cv_enable_compface=$enableval], [ac_cv_enable_compface=yes])
 if test "$ac_cv_enable_compface" = yes; then
-	AC_CHECK_LIB(compface, uncompface,,[ac_cv_enable_compface=no])
+	AC_CHECK_LIB(compface, uncompface,[],[ac_cv_enable_compface=no])
 fi
 
 dnl Check for libjconv
@@ -193,7 +191,7 @@
 	[  --disable-jconv         Do not use libjconv],
 	[ac_cv_enable_jconv=$enableval], [ac_cv_enable_jconv=yes])
 if test "$ac_cv_enable_jconv" = yes; then
-	AC_CHECK_LIB(jconv, jconv_alloc_conv,,[ac_cv_enable_jconv=no])
+	AC_CHECK_LIB(jconv, jconv_alloc_conv,[],[ac_cv_enable_jconv=no])
 fi
 
 AC_CHECK_LIB(xpg4, setlocale)
--- /dev/null	Tue May  5 22:32:27 1998
+++ sylpheed-0.7.8/acinclude.m4	Tue May  1 09:39:22 2001
@@ -0,0 +1,43 @@
+dnl SYLPHEED_CHECK_TYPE(TYPE, DEFAULT [, INCLUDES])
+dnl
+dnl Like AC_CHECK_TYPE, but in addition to `sys/types.h', `stdlib.h' and
+dnl `stddef.h' checks files included by INCLUDES, which should be a
+dnl series of #include statements.  If TYPE is not defined, define it
+dnl to DEFAULT.
+dnl
+AC_DEFUN([SYLPHEED_CHECK_TYPE],
+[AC_REQUIRE([AC_HEADER_STDC])dnl
+AC_MSG_CHECKING(for $1)
+AC_CACHE_VAL(sylpheed_cv_type_$1,
+[AC_TRY_COMPILE([
+#include <sys/types.h>
+#if STDC_HEADERS
+#include <stdlib.h>
+#include <stddef.h>
+#endif
+$3
+], [
+#undef $1
+int a = sizeof($1);
+], sylpheed_cv_type_$1=yes, sylpheed_cv_type_$1=no)])dnl
+AC_MSG_RESULT($sylpheed_cv_type_$1)
+if test $sylpheed_cv_type_$1 = no; then
+  AC_DEFINE($1, $2)
+fi
+])
+dnl GNUPG_CHECK_TYPEDEF(TYPE, HAVE_NAME)
+dnl Check whether a typedef exists and create a #define $2 if it exists
+dnl
+AC_DEFUN([GNUPG_CHECK_TYPEDEF],
+  [ AC_MSG_CHECKING(for $1 typedef)
+    AC_CACHE_VAL(gnupg_cv_typedef_$1,
+    [AC_TRY_COMPILE([#include <stdlib.h>
+    #include <sys/types.h>], [
+    #undef $1
+    int a = sizeof($1);
+    ], gnupg_cv_typedef_$1=yes, gnupg_cv_typedef_$1=no )])
+    AC_MSG_RESULT($gnupg_cv_typedef_$1)
+    if test "$gnupg_cv_typedef_$1" = yes; then
+        AC_DEFINE($2)
+    fi
+  ])
