diff -urN sylpheed-0.4.66.org/acconfig.h sylpheed-0.4.66/acconfig.h
--- sylpheed-0.4.66.org/acconfig.h	Sat May 12 12:11:34 2001
+++ sylpheed-0.4.66/acconfig.h	Sat May 12 12:44:21 2001
@@ -7,6 +7,8 @@
 #undef HAVE_GDK_PIXBUF
 #undef HAVE_GDK_IMLIB
 
+#undef HAVE_GTKHTML
+
 /* Whether to use multithread or not  */
 #undef USE_THREADS
 
diff -urN sylpheed-0.4.66.org/configure.in sylpheed-0.4.66/configure.in
--- sylpheed-0.4.66.org/configure.in	Sat May 12 12:11:34 2001
+++ sylpheed-0.4.66/configure.in	Sat May 12 12:48:58 2001
@@ -104,6 +106,28 @@
 	[ac_cv_disable_jconv=disable], [ac_cv_disable_jconv=no])
 if test "$ac_cv_disable_jconv" = no; then
 	AC_CHECK_LIB(jconv, jconv_alloc_conv)
+fi
+
+dnl Check for gtkhtml
+AC_ARG_ENABLE(gtkhtml,
+	[  --disable-gtkhtml         Do not use gtkhtml],
+	[ac_cv_disable_gtkhtml=disable], [ac_cv_disable_gtkhtml=no])
+AC_MSG_CHECKING([whether to use GtkHTML])
+if test "$ac_cv_disable_gtkhtml" = no; then
+	AC_MSG_RESULT(yes)
+	saved_CPPFLAGS="$CPPFLAGS"
+	CPPFLAGS="$CPPFLAGS $GTK_CFLAGS $GLIB_CFLAGS"
+	AC_CHECK_LIB(gtkhtml, gtk_html_new, ac_cv_gtkhtml_lib=yes, ac_cv_gtkhtml_lib=no, [ $GTK_LIBS $GLIB_LIBS  -lunicode ])
+	AC_CHECK_HEADER(gtkhtml/gtkhtml.h, ac_cv_gtkhtml_h=yes, ac_cv_gtkhtml_h=no)
+	if test "$ac_cv_gtkhtml_lib" = yes -a "$ac_cv_gtkhtml_h" = yes; then 
+		LIBS="$LIBS $GTK_LIBS $GLIB_LIBS -lgtkhtml -lunicode"
+		AC_DEFINE(HAVE_GTKHTML)
+	else
+	    	CPPFLAGS="$saved_CPPFLAGS"
+		AC_MSG_WARN([Unable to find GtkHTML])
+	fi
+else
+	AC_MSG_RESULT(no)
 fi
 
 AC_CHECK_LIB(xpg4, setlocale)
diff -urN sylpheed-0.4.66.org/src/Makefile.am sylpheed-0.4.66/src/Makefile.am
--- sylpheed-0.4.66.org/src/Makefile.am	Sat May 12 12:11:35 2001
+++ sylpheed-0.4.66/src/Makefile.am	Sat May 12 12:44:21 2001
@@ -11,6 +11,7 @@
 	headerview.c headerview.h \
 	textview.c textview.h \
 	imageview.c imageview.h \
+	htmlview.c htmlview.h \
 	mimeview.c mimeview.h \
 	summary_search.c summary_search.h \
 	folder.c folder.h \
diff -urN sylpheed-0.4.66.org/src/htmlview.c sylpheed-0.4.66/src/htmlview.c
--- sylpheed-0.4.66.org/src/htmlview.c	Thu Jan  1 01:00:00 1970
+++ sylpheed-0.4.66/src/htmlview.c	Sat May 12 12:44:21 2001
@@ -0,0 +1,117 @@
+/*
+ * Sylpheed -- a GTK+ based, lightweight, and fast e-mail client
+ * Copyright (C) 1999,2000 Hiroyuki Yamamoto
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <glib.h>
+#include <gtk/gtkscrolledwindow.h>
+#include <gtk/gtkpixmap.h>
+#include <gtk/gtktext.h>
+
+#ifdef HAVE_GTKHTML
+#  include <gtkhtml/gtkhtml.h>
+#endif /* HAVE_GTKHTML */
+
+#include "intl.h"
+#include "mainwindow.h"
+#include "prefs_common.h"
+#include "procmime.h"
+#include "htmlview.h"
+#include "utils.h"
+
+HTMLView *htmlview_create(void)
+{
+	HTMLView *htmlview;
+	GtkWidget *scrolledwin;
+	GtkWidget *html_widget;
+
+	debug_print(_("Creating html view...\n"));
+	htmlview = g_new0(HTMLView, 1);
+
+	scrolledwin = gtk_scrolled_window_new(NULL, NULL);
+	gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(scrolledwin),
+				       GTK_POLICY_NEVER,
+				       GTK_POLICY_ALWAYS);
+	gtk_widget_set_usize(scrolledwin, prefs_common.mainview_width, -1);
+
+#ifdef HAVE_GTKHTML
+	html_widget = gtk_html_new();
+	gtk_container_add (GTK_CONTAINER (scrolledwin), html_widget);
+#endif
+
+	gtk_widget_show_all(scrolledwin);
+
+	htmlview->scrolledwin  = scrolledwin;
+	htmlview->html_widget  = html_widget;
+	return htmlview;
+}
+
+void htmlview_init(HTMLView *htmlview)
+{
+}
+
+void htmlview_show_html(HTMLView *htmlview, MimeInfo *mimeinfo,
+			const gchar *file)
+{
+#ifdef HAVE_GTKHTML
+        GtkHTMLStream *handle;
+	GtkHTML *html;
+	FILE *fp;
+	char *buf, *c;
+	int bufsize;
+	const int maxbuf = 512; /* Read 512 bytes at a time */
+	
+	html = GTK_HTML(htmlview->html_widget);
+	if (!html) {
+		g_warning(_("Bad htmlview param."));
+		return;
+	} 
+	/* Open the file */
+	fp = fopen(file, "rt");
+	if (!fp) {
+		g_warning(_("Can't open HTML attachment."));
+		return;
+	}
+
+	buf = malloc(maxbuf);
+	if (!buf) {
+		g_warning(_("Can't allocate buffer for HTML."));
+		return;
+	}
+	
+	gtk_html_load_empty(html); 
+	handle = gtk_html_begin(html);
+	while (!feof(fp)) {
+	    bufsize = fread(buf, 1, maxbuf, fp);
+	    if (bufsize>0) {
+		gtk_html_write(html, handle, buf, bufsize);
+	    }
+	}
+	gtk_html_end(html, handle, GTK_HTML_STREAM_OK); 
+	free(buf);
+
+#endif
+}
+
+void htmlview_destroy(HTMLView *htmlview)
+{
+	g_free(htmlview);
+}
diff -urN sylpheed-0.4.66.org/src/htmlview.h sylpheed-0.4.66/src/htmlview.h
--- sylpheed-0.4.66.org/src/htmlview.h	Thu Jan  1 01:00:00 1970
+++ sylpheed-0.4.66/src/htmlview.h	Sat May 12 12:44:21 2001
@@ -0,0 +1,46 @@
+/*
+ * Sylpheed -- a GTK+ based, lightweight, and fast e-mail client
+ * Copyright (C) 1999,2000 Hiroyuki Yamamoto
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#ifndef __HTMLVIEW_H__
+#define __HTMLVIEW_H__
+
+#include <glib.h>
+#include <gtk/gtkwidget.h>
+
+typedef struct _HTMLView	HTMLView;
+
+#include "messageview.h"
+#include "procmime.h"
+
+struct _HTMLView
+{
+	GtkWidget *scrolledwin;
+	GtkWidget *html_widget;
+
+	MessageView *messageview;
+};
+
+HTMLView *htmlview_create	(void);
+void htmlview_init		(HTMLView	*htmlview);
+void htmlview_show_html		(HTMLView	*htmlview,
+				 MimeInfo	*mimeinfo,
+				 const gchar	*file);
+void htmlview_destroy		(HTMLView	*htmlview);
+
+#endif /* __HTMLVIEW_H__ */
diff -urN sylpheed-0.4.66.org/src/messageview.c sylpheed-0.4.66/src/messageview.c
--- sylpheed-0.4.66.org/src/messageview.c	Sat May 12 12:11:35 2001
+++ sylpheed-0.4.66/src/messageview.c	Sat May 12 12:44:21 2001
@@ -36,6 +36,7 @@
 #include "headerview.h"
 #include "textview.h"
 #include "imageview.h"
+#include "htmlview.h"
 #include "mimeview.h"
 #include "procmsg.h"
 #include "procheader.h"
@@ -62,6 +63,7 @@
 	HeaderView *headerview;
 	TextView *textview;
 	ImageView *imageview;
+	HTMLView *htmlview;
 	MimeView *mimeview;
 
 	debug_print(_("Creating message view...\n"));
@@ -77,9 +79,13 @@
 	imageview = imageview_create();
 	imageview->messageview = messageview;
 
+	htmlview = htmlview_create();
+	htmlview->messageview = messageview;
+
 	mimeview = mimeview_create();
 	mimeview->textview = textview;
 	mimeview->imageview = imageview;
+	mimeview->htmlview = htmlview;
 	mimeview->messageview = messageview;
 
 	vbox = gtk_vbox_new(FALSE, 0);
@@ -91,6 +97,7 @@
 	/* to remove without destroyed */
 	gtk_widget_ref(GTK_WIDGET_PTR(textview));
 	gtk_widget_ref(GTK_WIDGET_PTR(imageview));
+	gtk_widget_ref(GTK_WIDGET_PTR(htmlview));
 	gtk_widget_ref(GTK_WIDGET_PTR(mimeview));
 
 	messageview->vbox       = vbox;
@@ -99,6 +106,7 @@
 	messageview->headerview = headerview;
 	messageview->textview   = textview;
 	messageview->imageview  = imageview;
+	messageview->htmlview   = htmlview;
 	messageview->mimeview   = mimeview;
 
 	return messageview;
@@ -142,6 +150,7 @@
 	headerview_init(messageview->headerview);
 	textview_init(messageview->textview);
 	imageview_init(messageview->imageview);
+	htmlview_init(messageview->htmlview);
 	mimeview_init(messageview->mimeview);
 	//messageview_set_font(messageview);
 }
@@ -212,6 +221,7 @@
 {
 	TextView *textview = messageview->textview;
 	ImageView *imageview = messageview->imageview;
+	HTMLView *htmlview = messageview->htmlview;
 	MimeView *mimeview = messageview->mimeview;
 
 	if (messageview->type == type) return;
@@ -233,6 +243,9 @@
 		if (mimeview->vbox == GTK_WIDGET_PTR(textview)->parent) {
 			gtk_container_remove(GTK_CONTAINER(mimeview->vbox),
 					     GTK_WIDGET_PTR(textview));
+		} else if (mimeview->vbox == GTK_WIDGET_PTR(htmlview)->parent){
+			gtk_container_remove(GTK_CONTAINER(mimeview->vbox),
+					     GTK_WIDGET_PTR(htmlview));
 		} else {
 			gtk_container_remove(GTK_CONTAINER(mimeview->vbox),
 					     GTK_WIDGET_PTR(imageview));
@@ -257,17 +270,20 @@
 {
 	GtkWidget *textview  = GTK_WIDGET_PTR(messageview->textview);
 	GtkWidget *imageview = GTK_WIDGET_PTR(messageview->imageview);
+	GtkWidget *htmlview = GTK_WIDGET_PTR(messageview->htmlview);
 	GtkWidget *mimeview  = GTK_WIDGET_PTR(messageview->mimeview);
 
 	headerview_destroy(messageview->headerview);
 	textview_destroy(messageview->textview);
 	imageview_destroy(messageview->imageview);
+	htmlview_destroy(messageview->htmlview);
 	mimeview_destroy(messageview->mimeview);
 
 	g_free(messageview);
 
 	gtk_widget_unref(textview);
 	gtk_widget_unref(imageview);
+	gtk_widget_unref(htmlview);
 	gtk_widget_unref(mimeview);
 }
 
diff -urN sylpheed-0.4.66.org/src/messageview.h sylpheed-0.4.66/src/messageview.h
--- sylpheed-0.4.66.org/src/messageview.h	Sat May 12 12:11:35 2001
+++ sylpheed-0.4.66/src/messageview.h	Sat May 12 12:44:21 2001
@@ -29,6 +29,7 @@
 #include "headerview.h"
 #include "textview.h"
 #include "imageview.h"
+#include "htmlview.h"
 #include "mimeview.h"
 #include "procmsg.h"
 
@@ -49,6 +50,7 @@
 	HeaderView *headerview;
 	TextView *textview;
 	ImageView *imageview;
+	HTMLView *htmlview;
 	MimeView *mimeview;
 
 	MainWindow *mainwin;
diff -urN sylpheed-0.4.66.org/src/mimeview.c sylpheed-0.4.66/src/mimeview.c
--- sylpheed-0.4.66.org/src/mimeview.c	Sat May 12 12:11:35 2001
+++ sylpheed-0.4.66/src/mimeview.c	Sat May 12 12:44:21 2001
@@ -418,23 +418,64 @@
 	g_free(filename);
 }
 
+static void mimeview_show_html_part(MimeView *mimeview, MimeInfo *partinfo)
+{
+	gchar *name;
+	gchar *filename;
+
+	if (!partinfo) return;
+
+	name = partinfo->filename ? partinfo->filename
+		: partinfo->name ? partinfo->name : "tmpfile";
+	filename = g_strconcat(get_mime_tmp_dir(), G_DIR_SEPARATOR_S,
+			       name, NULL);
+
+	if (procmime_get_part(filename, mimeview->file, partinfo) < 0)
+		alertpanel_error
+			(_("Can't get the part of multipart message."));
+	else {
+		mimeview_change_view_type(mimeview, MIMEVIEW_HTML);
+		htmlview_show_html(mimeview->htmlview, partinfo, filename);
+	}
+
+	g_free(filename);
+}
+
 static void mimeview_change_view_type(MimeView *mimeview, MimeViewType type)
 {
 	TextView  *textview  = mimeview->textview;
 	ImageView *imageview = mimeview->imageview;
+	HTMLView *htmlview = mimeview->htmlview;
 
 	if (mimeview->type == type) return;
 
-	switch (type) {
+	switch(mimeview->type) {
 	case MIMEVIEW_IMAGE:
 		gtk_container_remove(GTK_CONTAINER(mimeview->vbox),
+				     GTK_WIDGET_PTR(imageview));
+		break;
+	case MIMEVIEW_HTML:
+  	        gtk_container_remove(GTK_CONTAINER(mimeview->vbox),
+				     GTK_WIDGET_PTR(htmlview));
+		break;
+	case MIMEVIEW_TEXT:
+		gtk_container_remove(GTK_CONTAINER(mimeview->vbox),
 				     GTK_WIDGET_PTR(textview));
+		break;
+	default:
+  	        return;
+	}
+
+	switch (type) {
+	case MIMEVIEW_IMAGE:
 		gtk_container_add(GTK_CONTAINER(mimeview->vbox),
 				  GTK_WIDGET_PTR(imageview));
 		break;
+	case MIMEVIEW_HTML:
+		gtk_container_add(GTK_CONTAINER(mimeview->vbox),
+				  GTK_WIDGET_PTR(htmlview));
+		break;
 	case MIMEVIEW_TEXT:
-		gtk_container_remove(GTK_CONTAINER(mimeview->vbox),
-				     GTK_WIDGET_PTR(imageview));
 		gtk_container_add(GTK_CONTAINER(mimeview->vbox),
 				  GTK_WIDGET_PTR(textview));
 		break;
@@ -475,7 +516,6 @@
 
 	switch (partinfo->mime_type) {
 	case MIME_TEXT:
-	case MIME_TEXT_HTML:
 	case MIME_MESSAGE_RFC822:
 	case MIME_MULTIPART:
 		mimeview_show_message_part(mimeview, partinfo);
@@ -485,6 +525,13 @@
 		mimeview_show_image_part(mimeview, partinfo);
 		break;
 #endif
+	case MIME_TEXT_HTML:
+#if (HAVE_GTKHTML)
+	        mimeview_show_html_part(mimeview, partinfo);
+#else
+		mimeview_show_message_part(mimeview, partinfo);
+#endif		
+		break;
 	default:
 		mimeview_change_view_type(mimeview, MIMEVIEW_TEXT);
 #if USE_GPGME
diff -urN sylpheed-0.4.66.org/src/mimeview.h sylpheed-0.4.66/src/mimeview.h
--- sylpheed-0.4.66.org/src/mimeview.h	Sat May 12 12:11:35 2001
+++ sylpheed-0.4.66/src/mimeview.h	Sat May 12 12:44:21 2001
@@ -28,13 +28,15 @@
 
 #include "textview.h"
 #include "imageview.h"
+#include "htmlview.h"
 #include "messageview.h"
 #include "procmime.h"
 
 typedef enum
 {
 	MIMEVIEW_TEXT,
-	MIMEVIEW_IMAGE
+	MIMEVIEW_IMAGE,
+	MIMEVIEW_HTML
 } MimeViewType;
 
 struct _MimeView
@@ -54,6 +56,7 @@
 
 	TextView *textview;
 	ImageView *imageview;
+	HTMLView *htmlview;
 
 	MessageView *messageview;
 
