diff -aurN sylpheed-2.1.4.orig/src/summaryview.c sylpheed-2.1.4/src/summaryview.c
--- sylpheed-2.1.4.orig/src/summaryview.c	2005-10-12 06:27:20.000000000 +0200
+++ sylpheed-2.1.4/src/summaryview.c	2005-10-23 15:12:15.000000000 +0200
@@ -4392,7 +4392,7 @@
 
 	gtk_tree_view_enable_model_drag_source
 		(GTK_TREE_VIEW(treeview),
-		 0, summary_drag_types, N_DRAG_TYPES,
+		 GDK_BUTTON1_MASK, summary_drag_types, N_DRAG_TYPES,
 		 GDK_ACTION_MOVE | GDK_ACTION_COPY);
 
 	g_signal_connect_after(G_OBJECT(treeview), "drag-begin",
