diff -Nru sylpheed-0.9.9-gtk2-20040229/configure.in sylpheed-0.9.9-gtk2-20040229.sam/configure.in
--- sylpheed-0.9.9-gtk2-20040229/configure.in	2004-02-23 11:56:36.000000000 +0000
+++ sylpheed-0.9.9-gtk2-20040229.sam/configure.in	2004-02-29 14:43:30.000000000 +0000
@@ -168,6 +168,20 @@
 	AC_DEFINE(HAVE_GDK_PIXBUF, 1, Define if you use gdk-pixbuf to support image view)
 fi
 
+dnl GtkSpell for spell checking in the compose window
+AC_ARG_ENABLE(gtkspell, [  --disable-gtkspell      Disable GtkSpell automatic spell checking [default=yes]],
+        [ac_cv_enable_gtkspell=$enableval], [ac_cv_enable_gtkspell=yes])
+AC_MSG_CHECKING([whether to use GtkSpell])
+if test "$ac_cv_enable_gtkspell" = yes ; then
+        AC_MSG_RESULT(yes)
+dnl FIXME this is probably not a clean way to do this
+        GTKSPELL_LIBS="$GTKSPELL_LIBS `pkg-config --libs gtkspell-2.0`"
+        AC_SUBST(GTKSPELL_LIBS)
+        AC_DEFINE(USE_GTKSPELL, 1, Define if you want GtkSpell to check spelling)
+else
+        AC_MSG_RESULT(no)
+fi
+
 dnl GPGME is used to support OpenPGP 
 AC_ARG_ENABLE(gpgme,
 	[  --enable-gpgme          Enable GnuPG support using GPGME [default=no]],
@@ -411,6 +425,7 @@
 else
 	echo "image support : no"
 fi
+echo "GtkSpell      : $ac_cv_enable_gtkspell"
 echo "GnuPG         : $ac_cv_enable_gpgme"
 echo "JPilot        : $ac_cv_enable_jpilot"
 echo "LDAP          : $ac_cv_enable_ldap"
diff -Nru sylpheed-0.9.9-gtk2-20040229/src/compose.c sylpheed-0.9.9-gtk2-20040229.sam/src/compose.c
--- sylpheed-0.9.9-gtk2-20040229/src/compose.c	2004-02-23 11:56:37.000000000 +0000
+++ sylpheed-0.9.9-gtk2-20040229.sam/src/compose.c	2004-02-29 14:43:30.000000000 +0000
@@ -4024,6 +4024,9 @@
 	gtk_widget_set_size_request(scrolledwin, prefs_common.compose_width, -1);
 
 	text = gtk_text_view_new();
+#ifdef USE_GTKSPELL
+        gtkspell_new_attach(GTK_TEXT_VIEW(text), NULL, NULL);
+#endif /* USE_GTKSPELL */
 	buffer = gtk_text_view_get_buffer(GTK_TEXT_VIEW(text));
 	gtk_text_view_set_editable(GTK_TEXT_VIEW(text), TRUE);
 	gtk_text_view_set_wrap_mode(GTK_TEXT_VIEW(text), GTK_WRAP_WORD);
diff -Nru sylpheed-0.9.9-gtk2-20040229/src/Makefile.am sylpheed-0.9.9-gtk2-20040229.sam/src/Makefile.am
--- sylpheed-0.9.9-gtk2-20040229/src/Makefile.am	2004-02-23 11:56:36.000000000 +0000
+++ sylpheed-0.9.9-gtk2-20040229.sam/src/Makefile.am	2004-02-29 14:43:30.000000000 +0000
@@ -200,6 +200,7 @@
 sylpheed_LDADD = \
 	@GTK_LIBS@ \
 	$(INTLLIBS) \
+	$(GTKSPELL_LIBS) \
 	$(GPGME_LIBS) \
 	$(LDAP_LIBS) \
 	$(LIBICONV) \
